### ISSUE TITLE
Deadlock when loading DataType classes

### ISSUE THREAD

**Issue by tstack** [jOOQ/jOOQ#3777]
We're seeing a deadlock when two different threads load different DataType classes at the same time.  It looks like loading one DataType class causes the others to be loaded as well during class initialization.  But, this can cause a deadlock when multiple threads are trying to load these DataType classes because they each hold a lock on the class and end up waiting for the others to finish.

The following code reproduces the deadlock for me:

``` java
import java.io.*;
import java.net.*;

class jooqlock {
    public static void main(String[] args) throws Exception {
        for (int i = 0; i < 1; i++) {
            doit();
        }
    }

    private static void doit() throws Exception {
        final ClassLoader loader = new URLClassLoader(new URL[] { new File("jooq-3.4.4.jar").toURI().toURL() });

        Runnable oload = new Runnable() {
                public void run() {
                    try {
                        Class.forName("org.jooq.util.sqlite.SQLiteDataType", true, loader);
                    } catch (Throwable th) {
                        th.printStackTrace();
                    }
                }
            };
        Runnable sload = new Runnable() {
                public void run() {
                    try {
                        Class.forName("org.jooq.util.postgres.PostgresDataType", true, loader);
                    } catch (Throwable th) {
                        th.printStackTrace();
                    }
                }
            };

        Thread th1 = new Thread(oload);
        Thread th2 = new Thread(sload);
        th1.start();
        th2.start();
        th1.join();
        th2.join();
    }
}
```

Here is a stack dump from after the deadlock happens:

```
Full thread dump Java HotSpot(TM) 64-Bit Server VM (24.65-b04 mixed mode):

"Thread-1" prio=5 tid=0x00007fd119841800 nid=0x5303 in Object.wait() [0x000000011b619000]
   java.lang.Thread.State: RUNNABLE
    at org.jooq.util.postgres.PostgresDataType.<clinit>(PostgresDataType.java:76)
    at java.lang.Class.forName0(Native Method)
    at java.lang.Class.forName(Class.java:270)
    at jooqlock$2.run(jooqlock.java:27)
    at java.lang.Thread.run(Thread.java:745)

"Thread-0" prio=5 tid=0x00007fd119841000 nid=0x5103 in Object.wait() [0x000000011b515000]
   java.lang.Thread.State: RUNNABLE
    at java.lang.Class.forName0(Native Method)
    at java.lang.Class.forName(Class.java:190)
    at org.jooq.impl.SQLDataType.<clinit>(SQLDataType.java:341)
    at java.lang.Class.forName0(Native Method)
    at java.lang.Class.forName(Class.java:190)
    at org.jooq.impl.DefaultDataType.<clinit>(DefaultDataType.java:219)
    at org.jooq.util.sqlite.SQLiteDataType.<clinit>(SQLiteDataType.java:71)
    at java.lang.Class.forName0(Native Method)
    at java.lang.Class.forName(Class.java:270)
    at jooqlock$1.run(jooqlock.java:18)
    at java.lang.Thread.run(Thread.java:745)
```

**Comment by gundeman** [comment]
We have also run into this issue. We worked around it by running this little code block on the applications main thread before starting up other threads:
>		try {
			Class.forName("org.jooq.impl.DefaultDataType");
		} catch (ClassNotFoundException e) {
			e.printStackTrace();
		}

**Comment by lukaseder** [comment]
@gundeman: Thank you very much for documenting this

**Comment by victorbr** [comment]
We happened to bump into this issue and while digging a bit into it I was wondering why the initialization lines in `SQLDataType` look like 
```
Class.forName(CUBRIDDataType.class.getName());
```
Unless I'm mistaken I believe it's an overkill - as soon as the code mentions `CUBRIDDataType.class`, it causes the class to be loaded. Looking it up again by name does nothing besides actually creating this deadlock.
I have tried replacing these lines by
```
Class<?> temp = CUBRIDDataType.class;
```
etc. and the deadlock no longer occurs.
@lukaseder do you think it could be a viable solution?

**Comment by lukaseder** [comment]
Thanks, @victorbr. Interesting thoughts. Since these deadlock occurred, there certainly had been quite a few changes to jOOQ's internals. I'll investigate your suggestion, soon.

**Comment by victorbr** [comment]
Just to clarify - this is easily reproducible on 3.10.0 with a test as simple as
```
    public static void main(String[] args) throws Exception {
        Runnable oload = () -> SQLDataType.VARCHAR(6);
        Runnable sload = () -> PostgresDataType.VARCHAR.length(6);
        Thread th1 = new Thread(oload);
        Thread th2 = new Thread(sload);
        th1.start();
        th2.start();
        th1.join();
        th2.join();
    }
```
(the original test in this thread, simplified)

**Comment by lukaseder** [comment]
Thanks for the additional heads-up

**Comment by lukaseder** [comment]
Yeah, well about that suggestion:

> Class<?> temp = CUBRIDDataType.class;

Try running this at the end of your main method:

```java
System.out.println(DSL.using(SQLDialect.MYSQL).render(cast(field("x"), SQLDataType.VARCHAR(6))));
```

Unfortunately, referencing the class doesn't necessarily load it. And if you did anything like referencing a member of the class, such as `CUBRIDDataType.VARCHAR`, you're back to the deadlock, as that would still have the same effect.

It's really a tricky problem to solve. I suspect it can only be solved like this: #5713, i.e. by getting rid of the internal static caches in `DefaultDataType`...

**Comment by victorbr** [comment]
> Try running this at the end of your main method:

Not sure I got this. With my version (referencing `.class`) it prints
```
cast(x as varchar(6))
```
With 3.10.0 it prints
```
cast(x as char(6))
```

**Comment by lukaseder** [comment]
Yeah, well the point is, `cast(x as varchar(6))` is illegal in MySQL. You can only cast to `char`. The dialects and their datatypes haven't been initialised correctly when you apply your fix.

**Comment by victorbr** [comment]
Aha! I see! Though I'm kinda puzzled by the current state - there's a clear circular dependency between `SQLDataType` and e.g. `MySQLDataType` static sections. I wonder how class loader manages to resolve it in the "normal" case...
I would love to help, just need to understand the intent of the current implementation! Appreciate your help!

**Comment by lukaseder** [comment]
Yes, there's a circular dependency, which isn't a problem if only one thread loads the classes:

- `SQLDataType` is loaded first, and with it, all standard data types are initialised
- When all `SQLDataType` *content* is initialised (but the class isn't completely loaded yet), then the dependencies (dialect-specific data types) are loaded. They can now refer to parent *content*, even if the class isn't completely initialised

This is a hack that persisted from early days on, and it works quite well in most cases. Every now and then, someone runs into this and the workaround is always to pre-initialise the data types.

This is not a nice situation of course, and it does need fixing.

**Comment by lukaseder** [comment]
We'll consider this issue fixed through the deprecation of the offending "internal" API: #7375. If it re-appears in other contexts, we can reopen this issue.

**Comment by etiennestuder** [comment]
@lukaseder I don't understand what the replacement for the deprecated class is, i.e. we are using org.jooq.util.postgres.PostgresDataType#TEXT, what should I replace this with in our code?

In general, it is really helpful when deprecating something to also mention what the new replacement is.

**Comment by etiennestuder** [comment]
Thanks for the insights, @lukaseder.

**Comment by raulvc** [comment]
sorry for bumping this dead post, any idea what's an equivalent for postgres DOUBLE PRECISION (float8) ?

**Comment by lukaseder** [comment]
@raulvc May I ask you to create a new issue with more details? https://github.com/jOOQ/jOOQ/issues/new/choose

**Comment by JoseLion** [comment]
@lukaseder I'm using JOOQ only for migrations (with Flyway), so I would like to have the exact PostgreSQL data types when I create  the tables (instead of the SQL equivalents). Just as an example, if I add a column with type `PostgresDataType.TEXT` it will be created as `text` on the database, but with `SQLDataType.VARCHAR` the type will be `varchar` in the database instead 😞 

If we want to keep PostgreSQL specific data types, what should we do? 🤔

**Comment by lukaseder** [comment]
> If we want to keep PostgreSQL specific data types, what should we do? 🤔

In general, please use a data type binding and attach that to the code generator's output:
https://www.jooq.org/doc/latest/manual/code-generation/custom-data-type-bindings

However, I think that it's currently not possible (without resorting to string manipulation) to override the data type literal when generating DDL statements.

**Comment by JoseLion** [comment]
I see 🤔 as I'm using it for migrations only, my best options is to go back to plain SQL files. However, thank you for the quick response on closed issue @lukaseder! 🙂
